### PR TITLE
#11: create ETL process for cucumber files

### LINKED ISSUE #11: Create ETL Process for Cucumber JSON Files
# Summary

RESO uses Cucumber JVM on the backend of testing, which generates both JSON and HTML reports. 

The HTML reports are large and are more for developers than normal users, so we're going to stop generating them and create our own reports from the JSON that gets generated instead. 

We will save the JSON reports in our versioned S3 buckets and process them in a Lambda function that uses this ETL package when certification fails.

The JSON reports are in Cucumber DSL, and we want to extract all errors and transform them into the format outlined here: https://github.com/RESOStandards/reso-certification/issues/1174

```
{
  "description": "RESO Certification Error Report",
  "endorsementName": "Data Dictionary",
  "version": "1.7",
  "generatedOn": "2022-10-12T06:14:07Z",
  "providerUoi": "<providerUoi>",
  "providerUsi": "<providerUsi>",
  "recipientEmail": "<primaryEmailAddress>",
  "certificationRequestId": "<certificationRequestId>",
  "status": "failed",
  "testResults": [
    { "testName": "Data Dictionary 1.7", "status": "passed" },
    {
      "testName": "Data Availability 1.7",
      "status": "failed",
      "errors": [
        {
          "stepName": "<name>",
          "message": "<errorMessage>"
        }
      ],
      "errorLog": [
         {
            "stepName": "<step name>",
            "message": "<error message>"
         }
    },
    { "testName": "IDX Payload 1.7", "status": "incomplete" }
  ]
}
```

The same format will be used for Data Dictionary, Data Availability, Web API Core, and any other endorsements RESO offers and the source structure will be the same (unless Cucumber JVM changes it) in all cases. We'll need to update this package if that occurs, but it's pretty stable and we can lock the versions. 

- [ ] We will pass in the step and status of items that failed, but we'll need to encode the different tests and their steps so we can fill in the rest. 

In the example above, we would pass the fact that it was Data Availability 1.7 that failed in a `failedStep` parameter and the ETL package should know that for Data Dictionary 1.7 IDX testing has three steps: Data Dictionary 1.7, Data Availability 1.7, and IDX Payload 1.7 and be able to fill in the rest.  

# Cucumber JSON Reports

Here is a sample report:
[data-dictionary-1.7.json.zip](https://github.com/RESOStandards/reso-certification-etl/files/9783589/data-dictionary-1.7.json.zip)

- [ ] We'll want to strip out all skipped and passed tests in this case and only process the errors. If there are none, then the process should return an empty `testResults` and `errorLog` entry. We want to make this flexible so we can potentially add skipped or passed tests in the future. 

The output is fairly hard to work with, but one of the inner errors is here: 

```
{
  "result": {
    "error_message": "java.lang.AssertionError: The following fields are missing the required 'RESO.OData.Metadata.LookupName' annotation: \nCountyOrParish, StreetDirSuffix, City, StateOrProvince, StreetDirPrefix, StreetSuffix\n\n\tat org.junit.Assert.fail(Assert.java:89)\n\tat org.reso.certification.stepdefs.LookupResource.resoLookupsUsingStringOrStringCollectionDataTypesMUSTHaveTheAnnotation(LookupResource.java:173)\n\tat ✽.RESO Lookups using String or String Collection data types MUST have the annotation \"RESO.OData.Metadata.LookupName\"(file:///home/jdarnell/work/reso/github/web-api-commander/src/main/java/org/reso/certification/features/data-dictionary/v1-7-0/additional-tests/lookup-resource-tests.feature:42)\n",
    "duration": 12000000,
    "status": "failed"
  },
  "line": 42,
  "name": "RESO Lookups using String or String Collection data types MUST have the annotation \"RESO.OData.Metadata.LookupName\"",
  "match": {
    "arguments": [
      { "val": "\"RESO.OData.Metadata.LookupName\"", "offset": 83 }
    ],
    "location": "org.reso.certification.stepdefs.LookupResource.resoLookupsUsingStringOrStringCollectionDataTypesMUSTHaveTheAnnotation(java.lang.String)"
  },
  "keyword": "Then "
}
```

In this case, we'd want to look for a status of `failed` for a given step. We want to capture both the `name` property as well as the `error_message` and `name` becomes `stepName` in the Cert API payload above, and `error_message` becomes `message`.

- [ ] The `error_message` needs additional formatting since it unfortunately contains JUnit Assertions as well. In this case, the following message:

```
"java.lang.AssertionError: The following fields are missing the required 'RESO.OData.Metadata.LookupName' annotation: \nCountyOrParish, StreetDirSuffix, City, StateOrProvince, StreetDirPrefix, StreetSuffix\n\n\tat org.junit.Assert.fail(Assert.java:89)\n\tat org.reso.certification.stepdefs.LookupResource.resoLookupsUsingStringOrStringCollectionDataTypesMUSTHaveTheAnnotation(LookupResource.java:173)\n\tat ✽.RESO Lookups using String or String Collection data types MUST have the annotation \"RESO.OData.Metadata.LookupName\"(file:///home/jdarnell/work/reso/github/web-api-commander/src/main/java/org/reso/certification/features/data-dictionary/v1-7-0/additional-tests/lookup-resource-tests.feature:42)\n"
```

would become:

```
"The following fields are missing the required 'RESO.OData.Metadata.LookupName' annotation: \nCountyOrParish, StreetDirSuffix, City, StateOrProvince, StreetDirPrefix, StreetSuffix"
```
In some cases this may be exactly the same message that's in the error.log, but not all errors are logged in the log that are logged in the Cucumber JSON steps, and vice versa, so we need to process both. 

We'll be refactoring the error messages to be more friendly with DD 2.0 so we can work on it more then. We'll probably use the error.log at that point rather than Cucumber JSON, in part to decouple our dependency on Cucumber JVM, and it would be easier if we could do that now but not every error is in the log. So, in the short term we'll need to parse the Cucumber JSON so we ship DD 1.7. 

# Optional Error Logs
- [ ] There may also be an error.log passed to the ETL function. If so, the ETL process should transform it into an array with one entry for each line. For example: 

```
[2022-10-06 21:13:40.566] [ERROR] [LookupResource] - ERROR: The following fields are missing the required 'RESO.OData.Metadata.LookupName' annotation: 
CountyOrParish, StreetDirSuffix, City, StateOrProvince, StreetDirPrefix, StreetSuffix
```

The single line above would be parsed as follows: 
```
{
 ...
  "errorLog": [
    {
      "testStep": "LookupResource",
      "message": "The following fields are missing the required 'RESO.OData.Metadata.LookupName' annotation: 
CountyOrParish, StreetDirSuffix, City, StateOrProvince, StreetDirPrefix, StreetSuffix" 
    }
  ]
}
```

Note that the timestamp and `[ERROR]` tags are parsed out, and if the message after the `-` character starts with "ERROR: " then it should be removed as well. Since it's raw text input, it's possible there could be nested quotes or double quotes in the message, so this should be handled appropriately. 

The error log format is shown here: https://github.com/RESOStandards/web-api-commander/blob/main/src/main/resources/log4j2.xml#L4. In the future, we may want to add timestamps as well but we'll want to  change them to ISO 8601 format.

### LINKED ISSUE #21: Create GitHub action to run ETL tests on PR
Workflow needed for running tests as part of CI.

### PR DESCRIPTION
**Pull request by mohit-s96**
fixes #11 

Adds ETL methods to process raw cucumber reports and to parse raw error logs. Also adds mocha tests for the same.

### CODE REVIEW

**Review by darnjo — CHANGES_REQUESTED** [review]
Looks good so far. 

Some things that are needed:
* Add Mocha test for error ETL process testing a passed-only file with the cucumber parser (should return nothing), an error-only file (should return only that error), and a file that has a passed and failed step (should return only the error and ignore passed).
* For error logs, the input will be multi line (each line separated with new lines). When this happens, the ETL process should have an array of errors, one for each message.

**Review by darnjo — CHANGES_REQUESTED** [review]
Need to change lib/reference-metadata/dd-1.7.json back to lib/references/dd-1.7/metadata-report.js

File: lib/common.js [hunk 1]
```diff
@@ -1,38 +1,42 @@
...
+const getStandardResources = version => {
+  const { resources = [] } = getMetadata(version);
+  return resources?.map(resource => resource.resourceName);
+};
 
 const getFieldDetails = async (fieldName, resourceName) => {
   return standardMetadata.fields.find(
-    (field) =>
-      field.fieldName === fieldName && field.resourceName === resourceName
+    field => field.fieldName === fieldName && field.resourceName === resourceName
```
Comment (darnjo) [inline lib/common.js:39]: Optional for `field?.fieldName` and `field?.resourceName`.

File: test/etl.js [hunk 2]
```diff
@@ -1 +1,106 @@
...
+      },
+      {
+        testStep: 'DataAvailability',
+        message:
+          'Request to https://retsapi.raprets.com/2/bari/RESO/OData/History?$filter=ModificationTimestamp lt 2022-12-19T22:21:05.499Z&$orderby=ModificationTimestamp desc&$top=100 failed! No response code was provided. Check commander.log for any errors...'
+      },
+      {
+        testStep: 'DataAvailability',
+        message:
+          'Request to https://retsapi.raprets.com/2/bari/RESO/OData/Offices?$filter=ModificationTimestamp lt 2022-12-19T22:23:08.573Z&$orderby=ModificationTimestamp desc&$top=100 failed! No response code was provided. Check commander.log for any errors...'
```
Comment (darnjo) [inline test/etl.js:95]: In general, we should try and remove the provider's API info. We can use something like test.reso.org.

**Review by darnjo — CHANGES_REQUESTED** [review]

**Review by darnjo — APPROVED** [review]

**Comment by mohit-s96** [comment]
A commit in this branch is being used on qa atm (00be7f490e06fb61bf0904703ed0791b59b736f0). We can delete this and move on to use the main branch once the frontend can support the new bins changes. cc:@darnjo

**Pull request merged**